src/celeritas/optical/model/RayleighModel.cc

```diff
@@ -0,0 +1,100 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/model/RayleighModel.cc
+//---------------------------------------------------------------------------//
+#include "RayleighModel.hh"
+
+#include "corecel/Assert.hh"
+#include "corecel/io/Logger.hh"
+#include "celeritas/io/ImportOpticalMaterial.hh"
+#include "celeritas/optical/MaterialParams.hh"
+#include "celeritas/optical/MfpBuilder.hh"
+
+#include "RayleighMfpCalculator.hh"
+
+namespace celeritas
+{
+namespace optical
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the model from imported data.
+ */
+RayleighModel::RayleighModel(
+    ActionId id,
+    SPConstImported imported,
+    SPConstMaterials materials,
+    std::vector<OpticalRayleighMaterial> rayleigh_materials)
+    : Model(id, "optical-rayleigh", "interact by optical Rayleigh")
+    , imported_(ImportModelClass::rayleigh, imported)
+    , materials_(std::move(materials))
+    , rayleigh_materials_(std::move(rayleigh_materials))
+{
+    CELER_EXPECT(materials_);
+    CELER_EXPECT(materials_->num_materials() == imported_.num_materials());
+    CELER_EXPECT(materials_->num_materials() == rayleigh_materials_.size());
+
+    for (auto mat : range(OpticalMaterialId(materials_->num_materials())))
+    {
+        CELER_EXPECT(imported_.mfp(mat) || rayleigh_materials_[mat.get()]);
```

Comment:
We should `VALIDATE` this input instead.

Reply:
Is it ok to do asserts in a loop like this, or should I have the whole check in something like
```
CELER_VALIDATE(std::all_of(...), << "...");
```
so the predicate is a single statement? I assume loops are fine I just haven't seen them for simple assertions in the code I've touched.

Reply:
`CELER_VALIDATE` is an always-on check, so this is fine. For assertions that are enabled with `CELERITAS_DEBUG` we can always do something like [this](https://github.com/celeritas-project/celeritas/blob/3269beff167334e28890838e7d9d9638341f015d/src/celeritas/em/model/detail/LivermoreXsInserter.hh#L67-L76).

Reply:
Yep, `CELER_VALIDATE` inside the loop is great for checking input data.